### PR TITLE
feat: [CCM-7278] Delegate Health Check

### PR DESCRIPTION
**Pull request by trunapushpa**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32919)

### CODE REVIEW

File: 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        clusterIds.add(clientId.substring(clientId.lastIndexOf('/') + 1));
+      } else {
+        clusterIds.add(perpetualTask.getClientContext().getClientParams().get("clusterId"));
+      }
+    }
+    Instant allowedTime = startTime.minus(Duration.ofMinutes(DELAY_IN_MINUTES_FOR_LAST_RECEIVED_MSG));
+    for (List<String> clusterIdsBatch : Lists.partition(clusterIds, BATCH_SIZE)) {
+      Map<String, Long> lastReceivedTimeForClusters =
+          lastReceivedPublishedMessageDao.getLastReceivedTimeForClusters(accountId, clusterIdsBatch);
+      for (String clusterId : lastReceivedTimeForClusters.keySet()) {
```
Comment (hitesharinga) [inline 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java:62]: We need to check if the delegate is healthy or not.

File: 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        if (Instant.ofEpochMilli(lastReceivedTimeForClusters.get(clusterId)).isBefore(allowedTime)) {
+          log.info("Delegate health check failed for clusterId: {}", clusterId);
```
Comment (hitesharinga) [inline 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java:64]: We need to send alerts for those clusters where we have't received the events at all/

File: scripts/terraform/stackdriver/batch-processing/delegate_health_monitoring.tf [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    "accountId" : "EXTRACT(jsonPayload.harness.accountId)",
+  }
+}
+
+resource "google_monitoring_alert_policy" "ce_delegate_health_monitoring_alert_policy" {
+  notification_channels = ((var.deployment == "prod" || var.deployment == "freemium" || var.deployment == "prod_failover") ? ["${local.slack_prod_channel}"] :
+    ((var.deployment == "qa" || var.deployment == "qa_free" || var.deployment == "stress") ? ["${local.slack_qa_channel}"] :
+  ["${local.slack_dev_channel}"]))
+
+  display_name = join("_", [local.name_prefix, "ce_delayed_batch_jobs"])
```
Comment (nikunj-badjatya) [inline scripts/terraform/stackdriver/batch-processing/delegate_health_monitoring.tf:29]: Change the display name to be unique.

**Review by nikunj-badjatya — CHANGES_REQUESTED** [review]

**Review by trunapushpa — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32919)* status: 8 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java` line 62 at r1](https://reviewable.io/reviews/harness/harness-core/32919#-N1bQ3JJRS92gtNApnkg-r1-62:-N1rDJ5869WrgOeoP_og:b-utawwg) ([raw file](https://github.com/harness/harness-core/blob/3c76978ccb49fe725c862edeb40fecdbd82e66a6/280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java#L62)):*
<details><summary><i>Previously, hitesharinga (Hitesh Aringa) wrote…</i></summary><blockquote>

We need to check if the delegate is healthy or not. 
</blockquote></details>

Done, but I couldn't test it as there were no healthy delegates on `ce-dev`.
___
*[`280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java` line 64 at r1](https://reviewable.io/reviews/harness/harness-core/32919#-N1bQ3JJRS92gtNApnkg-r1-64:-N1rGhvW6xq_NMJU2RoD:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/3c76978ccb49fe725c862edeb40fecdbd82e66a6/280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java#L64)):*
<details><summary><i>Previously, hitesharinga (Hitesh Aringa) wrote…</i></summary><blockquote>

We need to send alerts for those clusters where we have't received the events at all/ 
</blockquote></details>

Done.
___
*[`scripts/terraform/stackdriver/batch-processing/delegate_health_monitoring.tf` line 29 at r2](https://reviewable.io/reviews/harness/harness-core/32919#-N1ctTyZXgNpMsCYKyjC-r2-29:-N1rDQpZ0wMsCv-v5JUV:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/39fc164ad10d1440f112a9ae37721bf21b57a0e6/scripts/terraform/stackdriver/batch-processing/delegate_health_monitoring.tf#L29)):*
<details><summary><i>Previously, nikunj-badjatya (Nikunj Badjatya) wrote…</i></summary><blockquote>

Change the display name to be unique.
</blockquote></details>

Done.

File: 280-batch-processing/src/main/java/io/harness/batch/processing/ccm/BatchJobType.java [hunk 4]
```diff
@@ -40,6 +40,7 @@ public enum BatchJobType {
   AWS_ECS_SERVICE_RECOMMENDATION(130, 1, ChronoUnit.DAYS, singletonList(AWS_ECS_CLUSTER_DATA_SYNC), OUT_OF_CLUSTER_ECS),
   DEPLOYMENT_EVENT(150, 1, ChronoUnit.DAYS, emptyList(), IN_CLUSTER),
   K8S_EVENT(300, 1, ChronoUnit.HOURS, emptyList(), IN_CLUSTER),
+  DELEGATE_HEALTH_CHECK(310, 10, ChronoUnit.MINUTES, emptyList(), IN_CLUSTER),
```
Comment (hitesharinga) [inline 280-batch-processing/src/main/java/io/harness/batch/processing/ccm/BatchJobType.java:43]: Hey Truna, we have to update this also fetchLastBatchJobScheduledTime,
Change this to hourly right not its MINUTE. 
We should run it for only last 1 hr, What I mean is if job has not ran for 6 hrs so we don't have to run for all 6 hrs, just last 1 hr will be enough.  Like this
if (null != instant
        && ImmutableSet.of(BatchJobBucket.OUT_OF_CLUSTER, BatchJobBucket.OUT_OF_CLUSTER_ECS)
               .contains(batchJobType.getBatchJobBucket())
        && !ImmutableSet.of(BatchJobType.AWS_ECS_CLUSTER_SYNC, BatchJobType.AWS_ECS_SERVICE_RECOMMENDATION)
                .contains(batchJobType)) {
      Instant startInstant = Instant.now().minus(2, ChronoUnit.HOURS).truncatedTo(ChronoUnit.HOURS);
      instant = startInstant.isAfter(instant) ? startInstant : instant;
    }

File: 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    for (List<String> clusterIdsBatch : Lists.partition(clusterIds, BATCH_SIZE)) {
+      List<String> delegateIds =
+          clusterIdsBatch.stream().map(clusterIdToDelegateIdMap::get).collect(Collectors.toList());
+      List<Delegate> delegates = cloudToHarnessMappingService.obtainDelegateDetails(accountId, delegateIds);
+      Set<String> healthyDelegates = delegates.stream()
+                                         .filter(delegate -> isDelegateHealthy(delegate, startTime))
+                                         .map(Delegate::getUuid)
+                                         .collect(Collectors.toSet());
+      Map<String, Long> lastReceivedTimeForClusters =
+          lastReceivedPublishedMessageDao.getLastReceivedTimeForClusters(accountId, clusterIdsBatch);
```
Comment (hitesharinga) [inline 280-batch-processing/src/main/java/io/harness/batch/processing/tasklet/DelegateHealthCheckTasklet.java:81]: One optimisation we can do is get last received data for only healthy delegates.

File: 280-batch-processing/src/main/java/io/harness/batch/processing/ccm/BatchJobType.java (hunk 4 above)
Comment (hitesharinga) [reply 280-batch-processing/src/main/java/io/harness/batch/processing/ccm/BatchJobType.java:43]: Also first time it should run for last 1 hr. When you test on NG delete everything from BJSD for this job and then run.

**Pull request merged**